---
title: "eSCAN Package Tutorial"
author: Yingxi Yang
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{eSCAN Package Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction 

The `eSCAN` package accompanies the paper "eSCAN: Scan Regulatory Regions for Aggregate Association Testing using Whole Genome Sequencing Data" and is designed for performing genome-wide assessment of rare variants residing in enhancer regions that are significantly associated with a phenotype, combining the advantages of dynamic window selection with the advantages of increasing genomic annotation information. In this tutorial, we will use a toy example to show how to use eSCAN package.

# Setup Data

```{r warning=FALSE}
library(eSCAN)
data(geno) # genotype data frame
data(pheno) # phenotype and covariates data frame
data(enhancer) # enhancer data frame, optional
```
The genotype data in this tutorial is from the SCANG package, simulated using COSI, leveraging its calibrated model to closely resemble the LD patterns from African Americans. For the `pheno` data, it contains one column for phenotype and two for covariates which were generated from the standard normal distribution and Bernoulli distribution (p=0.5), respectively. The phenotype were then generated using the two covariates and error term, along with causal variants randomly selected from two signal regions (the signal regions were also randomly selected with length 3 kb). The enhancer data is as follows, generated by `eGenerator()` function in eSCAN package using parameter `maxgap=300`.
```{r}
head(enhancer)
```

The first column is index of the enhancers, the next two columns are start and end positions of the enhancers. The next two are start and end index of the enhancers (index in the genotype matrix sorted by genomic positions). The last column is the number of rare variants residing in the enhancers.

# Analyzing the simulated data using eSCAN

First, fit a generalized linear model under the null hypothesis. The parameter `x` is a data frame containing phenotype and covariates, each row indicates an individual and each column indicates a phenotype/covariate; the parameter `outcome` is used to specify the name of phenotype column in the data frame `x`, and the parameter `covars` is to specify the names of covariates. Then the parameter `fam` is either "gaussian" for a continuous phenotype or "binomial" for a dichotomous phenotype.

```{r}
nullmod <- fitNull(x = pheno, outcome = "phenotype", covars=c("X1", "X2"), fam="gaussian")
```

Then use the null model to run the main function `eSCAN()`. The parameter `times` indicates the number of Monte Carlo simulations (`default=1000`). For this toy example we set 10 so that you should be able to get results within one minute.

```{r}
res_eSCAN <- eSCAN(genotype = geno, nullmod = nullmod, new_enloc = enhancer, times=10, alpha=0.05, analy=FALSE)
```

The function returns a list containing 3 elements: `res`, `res0` and `thres`. `res` is a matrix of significant regions detected by eSCAN. The first column is the p-value of the detected region(s), and the next two columns are start and end position of the detected region(s).
```{r}
res_eSCAN$res
```

`res0` is a matrix to summarize all the regions included in the analysis.
```{r}
res_eSCAN$res0[1:10,]
```

`thres` is the threshold of eSCAN to control the family-wise/genome-wide error rate at `alpha` level.
```{r}
res_eSCAN$thres
```

